helix-core/src/syntax.rs

```diff
@@ -1809,56 +1887,44 @@ impl<'a> Iterator for HighlightIter<'a> {
-    for capture in query_match.captures {
-        let index = Some(capture.index);
-        if index == language_capture_index {
-            let name = byte_range_to_str(capture.node.byte_range(), source);
-            language_name = Some(name);
-        } else if index == content_capture_index {
-            content_node = Some(capture.node);
-        }
-    }
+#[derive(Debug, Clone)]
+pub enum InjectionCapture<'a> {
+    Name(Cow<'a, str>),
+    File(Cow<'a, Path>),
+    Shebang(String),
+}
 
-    let mut included_children = IncludedChildren::default();
-    for prop in query.property_settings(query_match.pattern_index) {
-        match prop.key.as_ref() {
-            // In addition to specifying the language name via the text of a
-            // captured node, it can also be hard-coded via a `#set!` predicate
-            // that sets the injection.language key.
-            "injection.language" => {
-                if language_name.is_none() {
-                    language_name = prop.value.as_ref().map(|s| s.as_ref().into())
-                }
-            }
+trait InjectionShebang<'a> {
+    fn injection_shebang(&self, source: &RopeSlice<'a>) -> Option<InjectionCapture<'a>>;
+}
 
-            // By default, injections do not include the *children* of an
-            // `injection.content` node - only the ranges that belong to the
-            // node itself. This can be changed using a `#set!` predicate that
-            // sets the `injection.include-children` key.
-            "injection.include-children" => included_children = IncludedChildren::All,
-
-            // Some queries might only exclude named children but include unnamed
-            // children in their `injection.content` node. This can be enabled using
-            // a `#set!` predicate that sets the `injection.include-unnamed-children` key.
-            "injection.include-unnamed-children" => included_children = IncludedChildren::Unnamed,
-            _ => {}
-        }
-    }
+impl<'a> InjectionShebang<'a> for Node<'a> {
+    fn injection_shebang(&self, source: &RopeSlice<'a>) -> Option<InjectionCapture<'a>> {
+        let node_slice = source.slice(self.byte_range());
+
+        // some languages allow space and newlines before the actual string content
+        // so a shebang could be on either the first or second line
+        let lines = if node_slice.len_lines() > 1 {
+            node_slice.slice(
+                // start of first line
+                node_slice.line_to_byte(0)
+                // end of second line
+                ..node_slice.line_to_byte(2) - 1,
+            )
+        } else {
+            node_slice.line(0)
+        };
+
+        static SHEBANG_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(SHEBANG).unwrap());
 
-    (language_name, content_node, included_children)
+        SHEBANG_REGEX
+            .captures(lines.as_str().unwrap_or_default())
+            .map(|cap| InjectionCapture::Shebang(cap[1].to_owned()))
+    }
```

Comment:
I don't think a trait and implementation are the right tool here. If anything I think this can be done in an ordinary function that takes a node and the rope as input

Reply:
How about an associated function on `InjectionCapture`? Just trying to avoid adding special purpose functions to the top-level scope.

Reply:
I don't think this function is so specific that it needs to be hidden. But the body is also pretty small so I think it would be ok to inline it into `injection_pair`.

As associated item feels awkward because it doesn't end up using `self` which I think is fine for functions like `new` or `default` but feels out-of-place in this usage

Reply:
I opted to just inline it in that case